### PR TITLE
[FIXME] Wrong results on main branch for INDF query

### PR DESCRIPTION
**Pull request by pobbatihari**
GPDB_12_MERGE_FIXME: the following comparison query has an ORCA plan that
relies on "IS NOT DISTINCT FROM" Hash Join, a variant that we likely have
lost during the merge with upstream Postgres 12. Disable ORCA for this query

Issue: HASH Join(left, right, inner) removing NULL tuples when there is
INDF in query.

The field hashqualclauses in HashJoin node holds the expressions "IS NOT
DISTINCT FROM" that test for match. This is normally idential to
hashclauses but differs in case of non-equijoin comparision.

While scanning the HashBucket we were using hashclauses so we missed
the expressions with "IS NOT DISTINCT FROM".

### CODE REVIEW

File: src/backend/executor/nodeHash.c [hunk 1]
```diff
@@ -2122,7 +2122,7 @@ bool
 ExecScanHashBucket(HashState *hashState, HashJoinState *hjstate,
 				   ExprContext *econtext)
 {
-	ExprState  *hjclauses = hjstate->hashclauses;
+	ExprState  *hjclauses = hjstate->hashqualclauses;
```
Comment (kainwen) [inline src/backend/executor/nodeHash.c:2125]: Let's add some gpdb specific comments here.

**Review by kainwen — APPROVED** [review]
LGTM.

File: src/test/regress/expected/select_distinct_optimizer.out [hunk 2]
```diff
@@ -334,7 +334,6 @@ SELECT null IS NOT DISTINCT FROM null as "yes";
 -- test IS DISTINCT FROM and IS NOT DISTINCT FROM join qual. the postgres planner doesn't support hash join
 -- on IS NOT DISTINCT FROM for now, ORCA support "IS NOT DISTINCT FROM" Hash Join but generates wrong result,
 -- DISABLE ORCA. Please fix me later if ORCA resolve the problem.
```
Comment (gpopt) [inline src/test/regress/expected/select_distinct_optimizer.out:336]: Can we remove this?

File: src/test/regress/expected/groupingsets_optimizer.out [hunk 3]
```diff
@@ -2428,16 +2426,14 @@ select g1000, g100, g10, sum(g::numeric), count(*), max(g::text) from
...
+NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause. Creating a NULL policy entry.
 set jit_above_cost to default;
 create table gs_hash_2 as
 select g1000, g100, g10, sum(g::numeric), count(*), max(g::text) from
   (select g/20 as g1000, g/200 as g100, g/2000 as g10, g
    from generate_series(0,19999) g) s
 group by cube (g1000,g100,g10);
-NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column(s) named 'g1000' as the Greenplum Database data distribution key for this table.
-HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
+NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause. Creating a NULL policy entry.
```
Comment (gpopt) [inline src/test/regress/expected/groupingsets_optimizer.out:2436]: Now Orca and planner generate table with different distribution policy.
Should we explicitly specify distribution key for CTAS table?

Comment (pobbatihari) [reply src/test/regress/expected/groupingsets_optimizer.out:2436]: yes, updated with distribution key. Thanks :)

File: src/backend/executor/nodeHash.c (hunk 1 above)
Comment (pobbatihari) [reply src/backend/executor/nodeHash.c:2129]: Added comments.

File: src/test/regress/sql/select_distinct.sql [hunk 4]
```diff
@@ -142,18 +142,31 @@ SELECT null IS NOT DISTINCT FROM null as "yes";
 -- test IS DISTINCT FROM and IS NOT DISTINCT FROM join qual. the postgres planner doesn't support hash join
 -- on IS NOT DISTINCT FROM for now, ORCA support "IS NOT DISTINCT FROM" Hash Join but generates wrong result,
 -- DISABLE ORCA. Please fix me later if ORCA resolve the problem.
```
[the remark this answers is not in the document]
Comment (pobbatihari) [reply src/test/regress/sql/select_distinct.sql]: updated Comments.

File: src/backend/executor/nodeHash.c [hunk 5]
```diff
@@ -2122,7 +2122,11 @@ bool
 ExecScanHashBucket(HashState *hashState, HashJoinState *hjstate,
 				   ExprContext *econtext)
 {
-	ExprState  *hjclauses = hjstate->hashclauses;
+	/*
```
Comment (kainwen) [inline src/backend/executor/nodeHash.c:2125]: Thanks for adding comments.  Minor suggestion:

for gpdb specific, we often encode special word like `CDB` or `Greenplum specific behavior` here.

File: src/backend/executor/nodeHash.c [hunk 6]
```diff
@@ -2122,7 +2122,12 @@ bool
...
+	 * Greenplum specific behavior.
+	 * Using hashqualclauses to support hash join on IS NOT DISTINCT FROM,
+	 * as well as '='.
+	 */
+	ExprState  *hjclauses = hjstate->hashqualclauses;
```
Comment (chrishajas) [inline src/backend/executor/nodeHash.c:2130]: In CTranslatorDXLToPlStmt, we have the code block
```
	if (!has_is_not_distinct_from_cond)
	{
		// no INDF conditions in the hash condition list
		hashjoin->hashclauses = hash_conditions_list;
	}
```

which populates hashclauses but not hashqualclauses. Will that cause an issue here?

File: src/test/regress/sql/select_distinct.sql [hunk 7]
```diff
@@ -139,9 +139,7 @@ SELECT 2 IS NOT DISTINCT FROM null as "no";
 SELECT null IS NOT DISTINCT FROM null as "yes";
 
 -- join cases
--- test IS DISTINCT FROM and IS NOT DISTINCT FROM join qual. the postgres planner doesn't support hash join
--- on IS NOT DISTINCT FROM for now, ORCA support "IS NOT DISTINCT FROM" Hash Join but generates wrong result,
--- DISABLE ORCA. Please fix me later if ORCA resolve the problem.
+-- test IS DISTINCT FROM and IS NOT DISTINCT FROM join qual.
```
Comment (gpopt) [inline src/test/regress/sql/select_distinct.sql:142]: Can we keep this part?
`the postgres planner doesn't support hash join on IS NOT DISTINCT FROM for now, ORCA support "IS NOT DISTINCT FROM" Hash Join.`

**Review by gpopt — APPROVED** [review]
LGTM after the comment is addressed.

**Review by chrishajas — APPROVED** [review]

File: src/test/regress/sql/select_distinct.sql (hunk 7 above)
Comment (pobbatihari) [reply src/test/regress/sql/select_distinct.sql:142]: ok

**Pull request merged**